internal/testing/runchecks.sh

```diff
@@ -43,6 +43,17 @@ else
   exit $result
 fi
 
+# Ensure that the code has no extra dependencies (including transitive
+# dependencies) that we're not already aware of by comparing with
+# ./internal/testing/alldeps
+#
+# Whenever a new valid dependency is added to the project, the PR should include
+# an updated ./internal/testing/alldeps
+# It's created by running "go list -deps ./... | sort"
+go list -deps ./... | sort | diff ./internal/testing/alldeps - || {
```

Comment:
I think this should be listing modules, not packages. Untested: I think you can get this with `-f '{{.Module.Path}}'`.

Also, subtractions are fine: additions should be what causes failure. I imagine something like `| sed -n -e '/^+/p'`, using `wc -l` to check for non-empty, and then echoing the output would be better.

Reply:
Great point about modules, thanks. I've updated the script now.

Re subtractions - I think we should, at least initially, be strict about it. If anyone removes a dependency, rerunning the script is not hard and helps us strictly keep track of dependencies. Otherwise if deps are not removed, they may be removed with an unrelated change much later on which will be confusing -- "why did this change remove dep XX, oh wait, it was removed long time ago, just wasn't updated in alldeps"

And having the two (output of script and alldeps) in perfect sync is more explicit and reason about.

WDYT?

Reply:
SGTM.